### DISCUSSION TITLE
Does ABS's m3u8 audio stream meet Amazon's criteria?

### DISCUSSION THREAD

**Discussion by amwmedia** [advplyr/audiobookshelf#3439] [Q&A]
I've started work to explore the feasibility of making an Alexa skill that interfaces with ABS. I've run into a roadblock that I'm trying to figure out. The skill plays the stream fine on my phone (through the Alexa app) but when I try to play it on a device I get nothing. Alexa support says to make sure the stream meets their url requirements. I believe it does, but was hoping someone on here might be able to point out something I missed?

[Their criteria](https://developer.amazon.com/en-US/docs/alexa/custom-skills/use-long-form-audio.html#url-requirements) are:
- You must host the audio file at an Internet-accessible HTTPS endpoint on port 443.
- The web server must present a valid and trusted SSL certificate. Self-signed certificates aren't allowed. Content hosting services, such as [Amazon S3](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html), provide valid and trusted SSL certificates.
- If the stream is a playlist container that references additional streams, you must host each stream within the playlist at an Internet-accessible HTTPS endpoint on port 443 with a valid and trusted SSL certificate.
- Your audio file must be in one of the following formats: AAC/MP4, MP3, PLS, M3U/M3U8, HLS.
- Your audio stream must support bit rates of 16 – 384 KB per second.

My test instance is hosted on a NAS and exposed to the public via Cloudflare tunnels so the SSL certificate shouldn't be an issue. The only criteria I am not sure about are the audio format used within the transport stream files.

**Comment by advplyr** [comment]
It should meet that criteria but the m3u8 stream is only returned when opening a session if it is requested. Depending on how you setup the skill would depend on what Abs sends back when opening the session.

**Reply by amwmedia** [reply]
Yeah, I have logic in the skill to make the POST request to start a new session and I pass the m3u8 `contentUrl` back to the echo device. It just doesn't want to play. I'm trying to get amazon to look at the stream url (I'm generating one and sending it with the `token` and all), but so far they have been less than helpful.

**Reply by advplyr** [reply]
I would suggest passing in `forceDirectPlay` when opening the session so you get back a url to the audio file. Unless you have audio formats in your library that aren't AAC/MP4 and MP3 you shouldn't need to transcode.

**Reply by amwmedia** [reply]
I've been primarily focused on `audioTracks[0].contentUrl` but I did try sending `libraryItem.media.tracks[0].contentUrl` as well with no luck. The latter seems to be a link to the (in my case anyway) raw m4b file. Is this what you're referring to? I have tried playing around with passing in things like `forceDirectPlay` or `forceTranscode` but I can't tell what changes about the payload that gets returned.

**Reply by advplyr** [reply]
The `audioTracks[0].contentUrl` is what changes when you `forceDirectPlay` or `forceTranscode`. Transcoding is the HLS m3u8 file.
When forcing direct play you may get many audioTracks depending on if you have a multi-file audiobook or not. It is up to the client to handle switching audio tracks.

**Reply by amwmedia** [reply]
Thanks for the clarification. We'll see if Amazon ever actually takes a look at this thing 🙄

**Comment by P-Hoep** [comment]
Oh man, yeah, running into this myself. It's pretty wild how poorly the audio player works, like, sometimes it will play through my dev test area, and not on devices, or on my phone and not on an actual echo device, etc. Kind of maddening.

**Reply by amwmedia** [reply]
Darn! I was hoping you had figured it out! 😉 

Yeah, and the worst part is I don't know what even needs to change to make it work. The stream seems fine and plays through something like VLC without any issues.

**Reply by amwmedia** [reply]
Have you even gotten the stream to play on a physical device?

**Reply by P-Hoep** [reply]
Yeah I can play it back on physical devices a few different ways (not .m3u8 though) edit - just saw you commented on this already, disregard if so). Playback works great on everything as best I can tell if you just grab the content url from the media tracks (it doesnt have a suffix .mp3 but seems to play fine anyway).

        if books:
            book = books[0]["libraryItem"]
            # Retrieve the MP3 file's contentUrl
            media_files = book['media']['tracks']
            logger.info(f"media_files {media_files}")
            if media_files:
                mp3_file = media_files[0]
                logger.info(f"mp3_file {mp3_file}")
                mp3_content_url = mp3_file.get('contentUrl')
                if mp3_content_url:
                    logger.info(f"mp3_content_url {mp3_content_url}")
                    full_mp3_url = f"{base_url}{mp3_content_url}?token={self.token}&t={timestamp}"
                    logger.info(f"MP3 URL: {full_mp3_url}")
                    self.currently_playing_url = full_mp3_url
                    return full_mp3_url
                else:
                    return "No content URL found for the MP3 file."
            else:
                return "No audio files found for the book."
                
 Apologies for the horribly messy code slamming my head against the wall on this one.
 
 The problem with that method though, is that its impossible to skip ahead or do anything like that (as best I can tell). The offset_in_milliseconds is just ignored with that stream. Which makes sense I guess, but boy would it be a lot easier if amazon let us fast forward/rewind a stream in realtime.
 
 I think that m3u8 is a relatively dead end - I tried downloading it in code, modifying it, uploading it to s3 with fully qualified urls and everything, no dice. Works just fine in VLC etc. though. However, on my physical devices the HLS streams extracted from the m3u8 file do play, so I'm sort of looking at pulling them out and then enqueueing them. Those streams don't play on my phone or in the dev test area though lol. I may do the above for fallback if there's a way to do isPlaying detection (no idea if thats possible or not though). 
 
 here's the extent I went through with the dumb m3u8 file. This will effectively create a valid new file with an accessible url with fully qualified urls for the .ts files but it still didn't work.
 
 import requests
import boto3
from botocore.exceptions import NoCredentialsError, ClientError
import uuid
import logging
from ask_sdk_model import ui
from utils import create_presigned_url

logger = logging.getLogger(__name__)
logger.setLevel(logging.INFO)

class M3U8Modifier:
    def __init__(self, base_url, m3u8_url, token):
        self.base_url = base_url
        self.m3u8_url = f"{m3u8_url}?token={token}"
        self.s3_client = boto3.client('s3')
        self.bucket_name = self.get_bucket()

    def get_bucket(self):
        response = self.s3_client.list_buckets()
        logger.info(f"buckets {response}")
        buckets = response['Buckets']
        if not buckets:
            raise Exception("No S3 buckets available")
        return buckets[0]['Name']

    def download_m3u8(self):
        logger.info(f"download url {self.m3u8_url}")
        response = requests.get(self.m3u8_url)
        if response.status_code != 200:
            raise Exception(f"Failed to download M3U8 file: {response.status_code}")
        return response.text

    def modify_m3u8(self, m3u8_content):
        lines = m3u8_content.splitlines()
        new_lines = []
        for line in lines:
            if ".ts" in line:
                new_lines.append(f"{self.base_url}{line}")
            else:
                new_lines.append(line)
        logger.info(f"newlines {new_lines}")
        return "\n".join(new_lines)

    def upload_to_s3(self, content):
        file_name = f"modified_{uuid.uuid4()}.m3u"
        logger.info(f"trying to upload https://{self.bucket_name}.s3.amazonaws.com/{file_name}")

        try:
            self.s3_client.put_object(
                Bucket=self.bucket_name,
                Key=file_name,
                Body=content,
                ContentType='application/vnd.apple.mpegurl'
            )
            logger.info(f"https://{self.bucket_name}.s3.amazonaws.com/{file_name}")
            
            # Generate presigned URL
            presigned_url = create_presigned_url(f"{file_name}")
            
            logger.info(f"presigned url {presigned_url}")
            return presigned_url
        except NoCredentialsError:
            raise Exception("AWS credentials not available.")

    def process_and_upload_m3u8(self):
        m3u8_content = self.download_m3u8()
        modified_m3u8 = self.modify_m3u8(m3u8_content)

        return self.upload_to_s3(modified_m3u8)

**Reply by P-Hoep** [reply]
If HLS enqueuing doesn't work out, I might go with trying to parse stuff out of the rss feed, but I haven't looked at that yet. https://github.com/sevenlayercookie/abs-alexa/blob/main/lambda/index.js

**Reply by P-Hoep** [reply]
Still no dice. came across this - I wonder if amazon is actively changing things. Eesh https://stackoverflow.com/questions/78767067/audio-stream-plays-on-alexa-app-but-not-on-echo-dot/78890427#78890427

**Reply by advplyr** [reply]
I'm still not sure why you are using the HLS stream instead of the mp3 or m4b file? Are you using a codec or audio container that is not supported by Alexa so you need to transcode it?

**Reply by amwmedia** [reply]
The reason I was preferring to use the HLS stream and ideally transcode is to standardize the output from whatever file type ADB has. Also because Alexa has very specific requirements and ADB may support (now or in the future) formats that Alexa doesn't support.

I did try passing the M4B link directly to Alexa and it choked on that too.

I received a reply from Amazon. They said that the link doesn't meet the requirement and that "The link must be public". I replied, asking what about the link is not "public"? The url is publicly accessible and only protected using a token query param (which, as far as I know, is no different from a pre-signed S3 link). I'll keep you guys updated when they reply.

**Reply by advplyr** [reply]
The HLS transcode is made on the fly which means the m3u8 playlist file is returned before all the segments are created. I first worked with HLS video a few years ago and there was no good resources I could find explaining on the fly streaming, so I did a lot of experimenting. https://github.com/advplyr/hls-media-server?tab=readme-ov-file#how-to-achieve-this

Some audio players will request segments far beyond the current time and it can be a problem if Abs is responding with too many 404s. You would see that in the server logs.

The HLS stream is a fallback option for when the client fails to direct play. Most audiobook libraries won't need to ever transcode.

**Reply by amwmedia** [reply]
Why would ABS respond with 404s? Just to clarify, are you saying ABS returns the m3u8 and then asynchronously starts generating the transport streams? Then if the client requests a bunch of the streams in parallel it might hit a bunch that don't exist yet?

**Reply by advplyr** [reply]
When calling the play API and getting back an m3u8 playlist file you should start seeing server logs about the progress of the transcode. When a request comes in that is asking for a segment that has not been created yet you will also see that in the logs. I'm not sure if you are looking at the server logs but that could give you some indication of what is going on

**Reply by amwmedia** [reply]
Just read over the link you posted, so I feel like I better understand what ABS is doing. I still don't get why it would be returning 404s.... unless.... Maybe the audio player in the echo devices has a VERY tight threshold for response time that isn't being met by ABS? Maybe?

Maybe I'll give another look at passing the m4b directly.

**Reply by advplyr** [reply]
404 is for Not Found because the requested segment does not exist yet

**Reply by amwmedia** [reply]
Ok, so it does work more like how I described. Based on your hls-media-server document I thought if a request comes in for stream 1, the server holds that request open while it generates the data and returns it. Then if a new request comes in for stream 10, the server would do the same thing.

**Reply by advplyr** [reply]
I don't understand what you mean. An HLS stream can be hundreds of files (segments) that take a while to generate. The m3u8 playlist file contains a list of URLs for every segment. Abs creates 6 second segments. It is the audio players job to request segments as they are needed, usually determined by the configured buffer amount.

Abs doesn't return anything when it finishes generating all the segments, it returns right away with the m3u8 playlist file. It starts generating segments based on the users progress. So if the user is 1 hour into the audiobook it starts a few segments before that.

**Reply by P-Hoep** [reply]
On my end when using the direct mp3 or m4a I wasn't able to get it to skip ahead chapters/minutes etc. - the Audioplayer has an offsetbymilliseconds parameter, which doesn't work at all with those types of data streams (I think). So the m3u8 file with the HLS chunks could have been a way to find an index to point at particular files relating to timestamps (although I never got that far, as it played like crap when I extracted the hls files from the m3u8 and enqueued them on the PlaybackFinishing handler). The Amazon Alexa skill Audioplayer is hot garbage lol.

**Reply by P-Hoep** [reply]
Playing the mp3 etc. directly though worked perfectly fine for me though, was very easy, but am stuck on playing from anywhere other than the very beginning of the file.

**Reply by P-Hoep** [reply]
I also tried converting the m3u8 to PLS, with no luck either. Tried various different header tags in the m3u8 file as well but without any luck on that either.

**Reply by P-Hoep** [reply]
if amazon insists on the audio urls being only publically available assets, I think over here they are triggering rss feed creation and parsing the files from there. So that could be a potential workaround. https://github.com/sevenlayercookie/abs-alexa

**Reply by advplyr** [reply]
There is a newer feature in Abs that isn't documented yet for creating media item shares. You can use the feature in the web client to get an idea of what the request/responses look like. The audio files are exposed publicly at `/share/:slug/track/:index`. You'll probably want to go through the codebase a bit if you want to use that method.

**Reply by amwmedia** [reply]
Would there be a way to map listening position back to a specific user? Sounds like there's no user-connected session to track progress?

**Reply by advplyr** [reply]
You would need to use the local session API endpoint and manage the session object yourself. It's probably not the route you want to go if you want to keep user progress synced.

**Reply by amwmedia** [reply]
It doesn't sound easy, but if the stream works in the audio player... It might be technically possible

**Reply by sevenlayercookie** [reply]
_One issue that I had making my skill (and the reason I went with RSS shares) is that Alexa devices do not handle cookies or auth headers. The echo device makes the request of the ABS API for the media file, but you cannot force the Echo device to include any headers or cookies with that request (which ABS requires). So you must be able to pass the echo device a plain, public media url, with no cookies or headers. 

There are two APIs for publicly accessible calls in ABS: media share, and RSS. Neither of these require auth header (so they're publicly accessible).

However, last I checked, media share item GET call **does** require including share_session_id cookie with the request, so this is fundamentally incompatible with Alexa. (However I'm not certain that share_session_id should be fundamentally required for the call, and I'm not certain what significant purpose the cookie serves anyways).

RSS shares do not require cookies or headers. I experimented with a man in the middle injection of headers and/or cookie using proxy, but RSS ended up being easier for me.

I also initially experimented with HLS and had issues (likely due to the same header and cookie issue however) and switched to direct play and have had perfect playback on all my varied Alexa devices with my skill, once I switched to RSS shares. It's possible HLS works if able to be done without headers or cookies, but I hadn't experimented further.

@advplyr if share_session_id cookie were optional when using the `GET: /api/share/:slug/track/:index` call, it would make Alexa playback a lot easier using media item share API. Perhaps if the API allowed the session_id to be passed as a URL parameter instead of a cookie. Or even better would be a new API for a dynamically generated temporary public URL that requires no authentication header or cookie.

In the meantime, I believe that RSS feed is the only option for Alexa (unless someone does MITM header or cookie injection, but I didn't love the security implications of that approach)_

EDIT: turns out, all of the RSS stuff is unnecessary. Alexa and ABS support streaming direct play **so long as the API key is included in the URL parameters**. I have updated my skill to use this, which greatly simplifies things….

**Reply by sevenlayercookie** [reply]
> It doesn't sound easy, but if the stream works in the audio player... It might be technically possible

It's possible; POC in my Alexa skill if you're interested ([https://github.com/sevenlayercookie/abs-alexa/](https://github.com/sevenlayercookie/abs-alexa/)). 

Since I used RSS feed approach, which has its own non-user-affiliated listening session, the workaround was creating a simultaneous user listening session in addition to the RSS listening session. Then the skill just manually syncs the two using the `POST http://abs.example.com/api/session/<ID>/sync` call. The API key you provide links the user session to the user that the API key is associated with, and updates that users current listening position and session duration stats etc.

**Comment by sevenlayercookie** [comment]
I have experimented more with Alexa skills and .m3u8 (unfortunately, very large audiobooks tend to crash the device using my ABS-Alexa skills, so I thought breaking it up using HLS might work better). However, I have not been able to get HLS to work on the echo devices, **regardless of source.** I’ve tried probably 20 different public HLS audio streams of various encodings on the internet (in addition to ABS HLS streams), and none of them play on the Alexa device. They do play on Alexa app and in the web browser.

so to answer your question, ABS is likely generating perfectly compatible streams; the problem is Alexa doesn’t actually support HLS, despite their website claiming otherwise.

**Reply by amwmedia** [reply]
Yeah, I have found that the main issue is the m3u8 file size. Most of the size comes from the token needing to be included with each transport stream. I've found that rewriting the m3u8 so that it doesn't need the token in the file generally brings the file size down to a level that the devices can handle.
